services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/DownloadFilter.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.function.BiPredicate;
+import software.amazon.awssdk.annotations.SdkPreviewApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+/**
+ * {@link DownloadFilter} allows you to filter out which objects should be downloaded as part of a {@link
+ * DownloadDirectoryRequest}. You could use it, for example, to only download objects of a given size, of a given file extension,
+ * of a given last-modified date, etc. Multiple {@link DownloadFilter}s can be composed together via {@link #and(BiPredicate)} and
+ * {@link #or(BiPredicate)} methods.
+ */
+@SdkPublicApi
+@SdkPreviewApi
+public interface DownloadFilter extends BiPredicate<S3Object, Path> {
```

Comment:
1. This extends `BiPredicate` so that we can directly take advantage of the default methods it offers, such as `and` and `or`.

2. I opted to call it `DownloadFilter`, as opposed to `DownloadDirectoryFilter`, since filtering would presumably only ever be relevant for batch-transfer-style operations.

Reply:
Hmm, `DownloadFileRequest` does capture more information (`TransferRequestOverrideConfiguration`, `GetObjectRequest`), but I'm not sure how likely that information is to be relevant for decision-making here. I feel like `<S3Object, Path>` is more symmetrical and elegant, and I'd prefer to keep that unless we have reason to believe users would need the additional information.

Reply:
Ack and mostly agree. However, there is one other small issue I just realized: `DownloadFileRequest` does not include `S3Object`, where the latter is needed for metadata like last modified time. So we would need to return a tuple/wrapper containing both `DownloadFileRequest` and `S3Object`, which feels quite cumbersome.

Reply:
Yeah, we'd have to use `BiPredicate<S3Object, DownloadFileRequest>` or Predicate<DownloadFileMetadata>`. Now I think about it, `DownloadFileMetadata` might be better. In the initial release, it can just have S3Object and Path, and we can add more if needed in the future.

Reply:
Oh, right, whoops! `DownloadFileRequest` has the `Path`. Realized that before but forgot.

Overall I'm still inclined towards `BiPredicate<S3Object, Path>` but I understand the argument towards `BiPredicate<S3Object, DownloadFileRequest>` or `Predicate<DownloadFileMetadata>`. Maybe this is a good one to let democracy decide.

Reply:
Going with `Predicate<DownloadFileMetadata>`!